codegen/gce/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+        let mut expressions_map = BTreeMap::new();
+
+        // maps names of named constants and public inputs to their NodeType
+        // Only CONST or VAR allowed
+        let mut const_public_type_map = BTreeMap::new();
+
+        // vector of expression nodes
+        let mut expressions = Vec::new();
+
+        // vector of `expressions` indexes
+        let mut outputs = Vec::new();
+
+        let num_polys = set_num_polys(ir, extension_degree);
+        let num_variables =
+            set_num_variables(ir, &mut const_public_type_map, boundary_constraints_vec);
+
+        // TODO #1: get rid of the vector and push values directly into result string
+        // constants from Constants AirIR field
+        // TODO #2: currently I add all found constants in the vector. Should I add only unique ones,
+        // since I'll get constants by their value, not index?
+        let constants = set_constants(ir, &mut const_public_type_map, boundary_constraints_vec);
+
+        set_transition_expressions(ir, &mut expressions, &constants, &mut expressions_map)?;
+        set_transition_outputs(ir, &mut outputs, &expressions_map)?;
+        set_boundary_expressions_and_outputs(
+            ir,
+            boundary_constraints_vec,
+            &mut expressions,
+            &mut outputs,
+            &constants,
+            &const_public_type_map,
+        )?;
+
+        Ok(GCECodeGenerator {
+            num_polys,
+            num_variables,
+            constants,
+            expressions,
+            outputs,
+        })
+    }
+
+    /// Generates constraint evaluation JSON file
+    pub fn generate(&self, path: &str) -> std::io::Result<()> {
+        let mut file = File::create(path)?;
+        file.write_all("{\n".as_bytes())?;
+        file.write_all(format!("\t\"num_polys\": {},\n", self.num_polys).as_bytes())?;
+        file.write_all(format!("\t\"num_variables\": {},\n", self.num_variables).as_bytes())?;
+        file.write_all(format!("\t\"constants\": {:?},\n", self.constants).as_bytes())?;
+        file.write_all(format!("\t\"expressions\": [\n\t\t{}", self.expressions[0]).as_bytes())?;
+        for expr in self.expressions.iter().skip(1) {
+            file.write_all(format!(",\n\t\t{}", expr).as_bytes())?;
+        }
+        file.write_all("\n\t],\n".as_bytes())?;
+        file.write_all(format!("\t\"outputs\": {:?}\n", self.outputs).as_bytes())?;
+
+        file.write_all("}\n".as_bytes())?;
+        Ok(())
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+fn set_num_polys(ir: &AirIR, extension_degree: u8) -> u16 {
+    // TODO: Should all aux columns be extended to be quadratic or cubic?
+    let num_polys_vec = ir.num_polys();
+    num_polys_vec
+        .iter()
+        .skip(1)
+        .fold(num_polys_vec[0], |acc, &x| {
+            acc + x * extension_degree as u16
+        })
+}
```

Comment:
Please add doc comments to this and all helper functions to explain the intended functionality (the meaning of the inputs & outputs)